// @coreui/react/dist/cjs/components/widgets/CWidgetStatsC.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var CCard = require('../card/CCard.js');
var CCardBody = require('../card/CCardBody.js');
require('../card/CCardFooter.js');
require('../card/CCardGroup.js');
require('../card/CCardHeader.js');
require('../card/CCardImage.js');
require('../card/CCardImageOverlay.js');
require('../card/CCardLink.js');
require('../card/CCardSubtitle.js');
require('../card/CCardText.js');
require('../card/CCardTitle.js');
var CProgress = require('../progress/CProgress.js');
var props = require('../../props.js');
var index = require('../../node_modules/classnames/index.js');

var CWidgetStatsC = React.forwardRef(function (_a, ref) {
    var className = _a.className, color = _a.color, icon = _a.icon, inverse = _a.inverse, progress = _a.progress, title = _a.title, value = _a.value, rest = tslib_es6.__rest(_a, ["className", "color", "icon", "inverse", "progress", "title", "value"]);
    return (React.createElement(CCard.CCard, tslib_es6.__assign({ className: className, color: color }, (inverse && { textColor: 'white' }), rest, { ref: ref }),
        React.createElement(CCardBody.CCardBody, null,
            icon && (React.createElement("div", { className: index.default('text-end mb-4', inverse ? 'text-white text-opacity-75' : 'text-body-secondary') }, icon)),
            value && React.createElement("div", { className: "fs-4 fw-semibold" }, value),
            title && (React.createElement("div", { className: inverse ? 'text-white text-opacity-75' : 'text-body-secondary' }, title)),
            React.createElement(CProgress.CProgress, tslib_es6.__assign({ className: "mt-3 mb-0", height: 4 }, (inverse && { white: true }), progress)))));
});
CWidgetStatsC.propTypes = {
    className: PropTypes.string,
    color: props.colorPropType,
    icon: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    inverse: PropTypes.bool,
    progress: PropTypes.object,
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.number]),
};
CWidgetStatsC.displayName = 'CWidgetStatsCWidgetStatsC';

exports.CWidgetStatsC = CWidgetStatsC;
