// @coreui/react/dist/cjs/components/card/CCard.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../node_modules/classnames/index.js');
var props = require('../../props.js');

var CCard = React.forwardRef(function (_a, ref) {
    var _b;
    var children = _a.children, className = _a.className, color = _a.color, textBgColor = _a.textBgColor, textColor = _a.textColor, rest = tslib_es6.__rest(_a, ["children", "className", "color", "textBgColor", "textColor"]);
    return (React.createElement("div", tslib_es6.__assign({ className: index.default('card', (_b = {},
            _b["bg-".concat(color)] = color,
            _b["text-".concat(textColor)] = textColor,
            _b["text-bg-".concat(textBgColor)] = textBgColor,
            _b), className) }, rest, { ref: ref }), children));
});
CCard.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    color: props.colorPropType,
    textBgColor: props.colorPropType,
    textColor: PropTypes.string,
};
CCard.displayName = 'CCard';

exports.CCard = CCard;
